### ISSUE TITLE
Use as an aligner

### ISSUE THREAD

**Issue by shangqwe123** [shivammehta007/Neural-HMM#9]
It is a great job!
I wonder if it can be used as an aligner. If so, the performance will be better than HMM-GMM or not.

**Comment by ghenter** [comment]
This is a neat idea! My answer will be "it almost can".

Right now, the code runs [the forward algorithm](https://en.wikipedia.org/wiki/Forward_algorithm) only, since that's all that's needed for maximum-likelihood training. However, one would need to run [the backward algorithm](https://en.wikipedia.org/wiki/Forward%E2%80%93backward_algorithm) or [the Viterbi algorithm (including the backwards pass)](https://en.wikipedia.org/wiki/Viterbi_algorithm) to obtain monotonic alignments. Work is currently being done to implement the Viterbi algorithm for our codebase.

I think the quality of the alignments obtained will depend on the training and test data. If you are trying to align to new, unseen speakers or if you only have small amounts of data (say half an hour or less) to train your system, I would recommend a GMM-HMM. Since neural HMM TTS uses a powerful neural decoder to process the input symbols (graphemes or phonemes), it can learn better pronunciation, which I imagine would help align uncommon and unusual words. On the other hand, this fancy decoder also gives the method more freedom, and it can learn things such as "let's use the last state of this word to describe the first sound in the word after it instead, since that gives lower training loss". This could make word boundaries less reliable. I would also expect many off-the-shelf aligners to be able to insert pauses between words where appropriate, something the current neural HMM TTS architecture does not do (it assumes the next word must start immediately after the previous one ends).

In summary, I think neural HMMs have the potential to be better aligners in certain scenarios, but it would require a bit of development to get there, and the code in this repo was neither developed nor evaluated with that goal in mind. My recommendation would be to stick with off-the-shelf GMM-HMMs if you just need a well-tested tool, but you could work on neural HMM aligners as a research problem, if that's what you want. In that case it would make sense to also look into monotonic latent-variable alignments for ASR, such as [this paper from RWTH Aaachen](https://arxiv.org/abs/2103.16710) (which also has [code available](https://github.com/rwth-i6/returnn-experiments/tree/master/2021-latent-monotonic-attention)).

I hope this answers your question!
